manpages/imodv.man
.nh
.na
.TH imodv 1 2.7.2 BL3DFS
.SH NAME
imodv \- View an imod model.
.SH SYNOPSIS
imodv [options] <model files...>
.SH DESCRIPTION
imodv is used to view models drawn with imod.  The main imodv window is
similar to the "Model View" window available in the program imod.  When
run from the command line, imodv can read and save models through the File 
menu.  Multiple models can be loaded and manipulated independently, either by 
putting their names on the command line or by loading them one after another
through the File menu.
The command line options are given below.
.SS Options
imodv uses all the standard Xt Toolkit options in addition to
the options described below.
.TP
.B -noborder
Open window with no border.
The menus can be accessed via a pop up menu using the right mouse button.
.TP
.B -f -fullscreen
Open model view window to maximum size supported by graphics.
No border will be created on the window.  
The menus can be accessed via a pop up menu using the right mouse button.
This mode is recommended if
one plans to use crystal eyes stereo hardware.
.TP
.B -rgb <color>  -renderbackground <color>
Select the color that imodv uses as the rendering background.
The color is a standard Xt Toolkit color name.
.TP
.B -cindex
Use color index mode. Useful on systems with only 8-bit graphics.
.TP
.B -D
Run in debug mode with diagnostic output and without becoming a background
process.
.P
.SS Hot key shortcuts for imodv.
.TP
.B - =
Zoom out/in with small steps.
.TP
.B _ +
Zoom out/in with large steps.
.TP
.B 1 2
Step back/forward in time.
.TP
.B 9 0 
Step to previous/next model.
.TP
.B 8
Toggle displaying all models or one model.
.TP
.B Arrow Keys, Page Up, Page Down
Translate model in X, Y, and Z.
.TP
.B Numeric Keypad
Model rotation controls:
   The 4 and 6 keys rotate about the Y Axis. 
   The 8 and 2 keys rotate about the X Axis.
   The 9 and 3 keys rotate about the Z Axis.  
   The 5 Key toggles between movie and incremental motion.
.TP
.B , .
Decrease or increase the rotation increment and movie speed.

.TP
.B Escape, q
Quit imodv or close model view window.
.TP
.B B
Open dialog to set background color.
.TP
.B C
Open the Control editing dialog.
.TP
.B i
Open the Image overlay control dialog.
.TP
.B L
Open Object List dialog.
.TP
.B m
Open Movie control dialog.
.TP
.B M
Open Model editing dialog.
.TP
.B O
Open the Object editing dialog.
.TP
.B v
Open the View editing dialog.
.TP
.B S
Snapshot image
.TP
.B s
Toggle Stereo On/Off.
.TP
.B [ ]
Adjust parallax for stereo viewing.
.TP
.B l
Invert the parallax angle.
.TP
.B b
Toggle single/double buffering.
.TP
.B r
Toggle low resolution drawing of mesh and spheres.
.TP
.B g G
Increase or decrease the quality (and time) of sphere drawing.
.TP
.B c
Print clip plane information to standard output.
.TP
.B o
Print model transformations to standard output.

.P
Use the online help for more information on using imodv.
.SH AUTHORS
.nf
Jim Kremer
David Mastronarde
.fi
.SH SEE ALSO
imod(1), imodinfo(1)
.SH BUGS
Please email all bug reports to mast@colorado.edu.


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
10/8/03: Modified PIP to read newly invented "autodoc" files to determine
program options and generate manual pages sections.

10/6/03: IMOD_3-0-22
10/6/03: Fixed the default number of views value for transferfid; it was not
being initialized in etomo.

10/6/03: Reordered thread control to eliminate race condition that often 
prevented matchvol1 from running right after any preceeding com script.

10/2/03: IMOD_3-0-21
10/2/03: Modified Rotatevol, Matchvol, and Warpvol to work on volumes up to 
16K in any dimension; Rotatevol was failing for an output X/Y size of 4K by 4K.

10/1/03: Made imodinfo work properly for counting scattered points on subsets
of the volume.

10/1/03: Added message for opening bead fixer window.

10/1/03: Added ability to turn a plugin into an internal module that is added
to the Special menu, to provide Bead fixer on Windows.

9/29/03: Restructured some 3dmod include files to restore ability to build
plugins with minimal header files.

9/25/03: Made sure windows stayed on the screen when restoring Zap and Info
window positions.

9/25/03: Changed Midas to number sections from 1 instead of 0.

9/24/03: IMOD_3-0-20
9/24/03: Resolved problems with the Fortran graphics package so it works
on all systems, at least briefly.

9/24/03: Fixed some problems with event sending under Windows, added a
retry capability to make it work reliably on the Mac, and found that it
did not work with the multi-threaded Qt library on the SGI.

9/24/03: Used different window properties for saving and restoring window
positions, which fixed problems in Redhat 7.3 and Irix.  Made it not restore
Info window size since that is highly font-dependent.

9/23/03: Put Fortran graphics programs inside application directories on the
Mac, which made them work correctly, with the window raisable and resizable.

9/22/03: Removed a large unused common from smoothsurf, which was making it
fail under Cygwin.

9/19/03: Made sample.com files produce top.rec, etc, directly as the output
of Tilt to avoid replacing existing .rec files.

9/19/03: Made 3dmod require only a 15-bit RGB visual, not 16 bit.

9/18/03: IMOD_3-0-19
9/18/03: Added the ability to float (match contrast between sections) based
on a subarea, either the image displayed or the area inside of a rubberband
in the active Zap window.  Also added an autocontrast button to the Info
window with target contrasts settable from the options dialog.

9/17/03: Made the Zap window pan by one display pixel per pixel of mouse move
for zooms less than one.

9/16/03: Made 3dmod save and restore sizes and positions of the Zap and Info
windows, with different settings saved for different image sizes.  Zoom is
adjusted to fit the whole image to the window.  If there are no application
saved settings, the Zap window now opens up wit a size that fits a zoomed-down
image if appropriate.

9/15/03: Eliminated physical flipping of data in 3dmod to avoid very bad
performance when the image data gets into swap space.

9/15/03: Optimized image drawing for zooms equal to the inverse of an integer.

9/13/03: IMOD_3-0-18
9/12/03: Prevented buffer overruns in an array that holds the filename in
3dmod, fixed problems with model checksums that were leading to queries to
save an unchanged model, arranged the startup page in two columns, fixed a
problem with zooming up under Windows, and added a workaround to a display
problem with Qt 3.2.1 on the Mac.

9/11/03: IMOD_3-0-17  beta release with Etomo

9/9/03: IMOD_3-0-16
9/5/03: Added a program, Resamplemod, for rotating a microtubule model to be
perpendicular to the Z axis and creating points at each section or plane
along that axis.

9/4/03: Incorporated changes in Mtlengths to correct for oblique microtubules
in tilted sections and implemented PIP input.

8/29/03: Set up and documented installation procedures on the Mac.

8/29/03: Changed the Fortran graphics routines to run the graphics window in
a separate thread from the Fortran program.  This only works properly under
Linux, where the window refreshes itself when resized or exposed.  On the Mac,
the window can now be moved but not resized.

8/25/03: Added the -E option to Imodmesh to cap the ends of tubes.

8/13/03: IMOD_3-0-15
8/12/03: Made the Qt-based graphics routines used in Fortran programs work on
the Mac without hanging when the window opened.

8/12/03: Made it possible to start up Nda without a model file, just like Mtk.

8/9/03: IMOD_3-0-14
8/9/03: Changed Blendmont to take PIP input and modified command files and
Copytomocoms to use new form of input.

8/8/03: Etomo was incorporated into IMOD build structure.

8/8/03: Wrote man page for Mtk, changed program to make terminology
consistent, to fix problems with saving connector objects and moving items to
new objects, and to add ability to export a graph in a format easily readable
by other programs.  (Nda can also export graphs.)  Also provided some example
command files in the com directory.

8/8/03: Fixed problem of current point not being displayed when model is
turned off, caused by the ability to display the vector from bead fixer.

8/1/03: Made it possible to kill 3dmod during image flipping operation.

8/1/03: Standardized error exit in Xfproduct.

8/1/03: Implemented proper thickness padding in fast backprojection code of
Tilt, reduced Tilt's stack size so the increase is only 3-fold from previously,
and made memory usage and data loading dynamic so that it will not use more
than 80 MB of stack unless it needs to.

7/31/03: IMOD_3-0-13
7/31/03: Added ability to send messages to set movie mode and to load a model
without changing black/white levels.

7/31/03: Changed the call used to iconify dialog boxes along with the info
or model view window because the old way didn't work under RedHat 9 or on the 
SGI under Qt 3.1.

7/31/03: Fixed remaining problems created by new treatment of "views", and

imod/imodv.cpp
/*  IMOD VERSION 2.50
 *
 *  imodv.cpp -- The main imodv entry point for standalone or imod operation
 *
 *  Original author: James Kremer
 *  Revised by: David Mastronarde   email: mast@colorado.edu
 */

/*****************************************************************************
 *   Copyright (C) 1995-2001 by Boulder Laboratory for 3-Dimensional Fine    *
 *   Structure ("BL3DFS") and the Regents of the University of Colorado.     *
 *                                                                           *
 *   BL3DFS reserves the exclusive rights of preparing derivative works,     *
 *   distributing copies for sale, lease or lending and displaying this      *
 *   software and documentation.                                             *
 *   Users may reproduce the software and documentation as long as the       *
 *   copyright notice and other notices are preserved.                       *
 *   Neither the software nor the documentation may be distributed for       *
 *   profit, either in original form or in derivative works.                 *
 *                                                                           *
 *   THIS SOFTWARE AND/OR DOCUMENTATION IS PROVIDED WITH NO WARRANTY,        *
 *   EXPRESS OR IMPLIED, INCLUDING, WITHOUT LIMITATION, WARRANTY OF          *
 *   MERCHANTABILITY AND WARRANTY OF FITNESS FOR A PARTICULAR PURPOSE.       *
 *                                                                           *
 *   This work is supported by NIH biotechnology grant #RR00592,             *
 *   for the Boulder Laboratory for 3-Dimensional Fine Structure.            *
 *   University of Colorado, MCDB Box 347, Boulder, CO 80309                 *
 *****************************************************************************/

/*  $Author$

$Date$

$Revision$
Log at end of file
*/

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <math.h>
#include "imodv_window.h"
#include <qapplication.h>
#include <qdir.h>
#include "dia_qtutils.h"

#include "imodv.h"
#include "imod.h"
#include "imod_object_edit.h"
#include "imod_display.h"
#include "imodv_gfx.h"
#include "imodv_stereo.h"
#include "imodv_modeled.h"
#include "preferences.h"

// static declarations
static void usage(char *pname);
static int load_models(int n, char **fname, ImodvApp *a);
static int openWindow(ImodvApp *a);
static int getVisuals(ImodvApp *a);
static void initstruct(ImodView *vw, ImodvApp *a);
static int imodv_init(ImodvApp *a, struct Mod_Draw *md);


// GLOBAL VARIABLES for the imodv and draw structures
ImodvApp  ImodvStruct;
struct Mod_Draw Imodv_mdraw;
ImodvApp *Imodv = &ImodvStruct;

/* A global to indicate if the window is closed */
int ImodvClosed = 1;


#define DEFAULT_XSIZE  512
#define DEFAULT_YSIZE  512


/* the default graphics rendering attributes for OpenGL */
static ImodGLRequest True24 = {1, 1, 24, 1, 1};
static ImodGLRequest True15 = {1, 1, 15, 1, 1};
static ImodGLRequest True12 = {1, 1, 12, 1, 1};
static ImodGLRequest True24nodep = {1, 1, 24, 0, 1};
static ImodGLRequest True15nodep = {1, 1, 15, 0, 1};
static ImodGLRequest True12nodep = {1, 1, 12, 0, 1};

/* List the attribute lists in order of priority */
static ImodGLRequest *OpenGLAttribList[] = {
  &True24, &True15, &True12, &True24nodep, &True15nodep, &True12nodep, 
  NULL
};

/*void __eprintf(void){return;} */

static void usage(char *pname)
{
  imodVersion(pname);
  imodCopyright();
  fprintf(stderr, "options: all Qt options plus:\n");
  fprintf(stderr, "\t-f                Open window to max size.\n");
  fprintf(stderr, "\t-b color_name     Background color for rendering.\n");
  fprintf(stderr, "\t-s width,height   Window size in pixels.\n");
  fprintf(stderr, "\t-D                Debug mode.\n");
  exit(-1);
}

static char *blackString = "black";

/* DEFAULT INITIALIZATION OF THE STRUCTURES
   9/2/02: also called for model view initialization in imod */
static int imodv_init(ImodvApp *a, struct Mod_Draw *md)
{
  a->nm = 0;
  a->cm = 0;
  a->mod = NULL;
  a->imod = NULL;
  a->mat  = imodMatNew(3);
  a->rmat  = imodMatNew(3);
  a->dobj = imodObjectNew();
  a->obj = a->dobj;
  a->ob = 0;
  a->md = md;
  a->cnear = 0;
  a->cfar = 1000;
  a->fovy = 0;
  a->dlist = 0;
  a->update_dlist = 1;
  a->movie = 0;
  a->movieFrames = 0;
  a->wpid = 0;
  a->stereo = IMODV_STEREO_OFF;
  a->plax = 5.0f;
  a->lightx = a->lighty = 0;
  a->winx = DEFAULT_XSIZE;
  a->winy = DEFAULT_YSIZE;
  a->rbgname = blackString;
  a->rbgcolor = new QColor(0, 0, 0);

  md->xorg = md->yorg = md->zorg = 0;
  md->xrot = md->yrot = md->zrot = 0;  /* current rotation. */
  md->zoom = 1.0f;
  md->arot = 10;
  md->atrans = 5.0f;
  md->azoom = 1.05f;
  md->azscale = 0.2;
  md->xrotm = 0; /* rotation movie */
  md->yrotm = 0;
  md->zrotm = 0;

  /* control flags */
  a->fastdraw   = 0;
  a->current_subset = 0;
  a->crosset    = 0;
  a->fullscreen = 0;
  a->drawall    = 0;
  a->moveall    = 1;
  a->alpha      = 0;
  imodViewDefault(&a->view);
  a->view.cnear = 0.0f;
  a->view.cfar  = 1.0f;
  a->doPick = 0;
  a->wPick = a->hPick = 5;

  a->lighting  = 1;
  a->depthcue  = 0;
  a->wireframe = 0;
  a->lowres = 0;
  a->SGIStereoCommand = NULL;
  a->SGIRestoreCommand = NULL;
  return(0);
}

// ADDITIONAL INITIALIZATION TO OPEN MODEL VIEW FROM IMOD
/* DNM 9/2/02: changed to call imodv_init for bulk of the initialization */
static void initstruct(ImodView *vw, ImodvApp *a)
{

  imodv_init(a, &Imodv_mdraw);

  a->nm = 1;
  a->mod = (Imod **)malloc(sizeof(Imod *));
  a->mod[0] = vw->imod;
  a->imod = vw->imod;

  /* DNM 8/3/01: start with current object if defined */
  if (a->imod->cindex.object >= 0 && 
      a->imod->cindex.object < a->imod->objsize) {
    a->ob = a->imod->cindex.object;
    a->obj = &(a->imod->obj[a->ob]);
  }

  /* control flags */
  a->fullscreen = 0;

  a->standalone = 0;
  a->texMap  = 0;
  a->texTrans = 0;
  a->vi = vw;

  a->SGIStereoCommand  = ImodRes_SGIStereoCommand();
  a->SGIRestoreCommand = ImodRes_SGIRestoreCommand();

  imodViewStore(a->imod, 0);
  if (!a->imod->cview)
    imodViewModelDefault(a->imod, a->imod->view);
  else {
    imodViewUse(a->imod);
    imodvDrawImodImages();
  }
  return;
}


// GET THE VISUAL INFORMATION NEEDED TO OPEN GL WIDGETS
/* DMN 9/2/02: replaced old imodvSetCmap with this, which is accessed from
   both imodv and ximodv (model view startup in imod), and which gets the
   best visuals for double and single buffering then does the old work of
   imodvSetCmap to get color map if needed in color index mode */

static int getVisuals(ImodvApp *a)
{
  int i, colorDB, colorSB;
  int depthSB = -1;
  int depthDB = -1;
  ImodGLVisual *visual;

  // Keep track of what depth request is used to get the visuals selected
  // and of the actual depth bits that result
  a->enableDepthSB = -1;
  a->enableDepthDB = -1;

  // Loop through all requests, asking first for double buffer then for single
  // buffer.  When one is found, the depth is set
  for (i = 0; OpenGLAttribList[i] != NULL; i++) {
    if (depthDB < 0) {
      visual = imodFindGLVisual(*OpenGLAttribList[i]);
      if (visual) {
        a->enableDepthDB = visual->depthEnabled;
	a->stereoDB = visual->stereo;
        depthDB = visual->depthBits;
        colorDB = visual->colorBits;
      }
    }
      
    if (depthSB < 0) {
      OpenGLAttribList[i]->doubleBuffer = 0;
      visual = imodFindGLVisual(*OpenGLAttribList[i]);
      if (visual) {
        a->enableDepthSB = visual->depthEnabled;
	a->stereoSB = visual->stereo;
        depthSB = visual->depthBits;
        colorSB = visual->colorBits;
      }
    }

    /* If got both, stop the loop */
    if (depthDB >= 0 && depthSB >= 0)
      break;
  }

  /* error if no visuals */
  if (depthDB < 0 && depthSB < 0)
    return 1;

  if (!depthDB || !depthSB)
    fprintf(stderr, "Imodv warning: using a visual with"
            " no depth buffer\n");

  if (depthDB < 0)
    fprintf(stderr, "Imodv warning: no double buffer visual available.\n");
  else if (Imod_debug)
    printf("DB visual: %d color bits, %d depth bits, stereo %d\n",
	   colorDB, depthDB, a->stereoDB);
  if (depthSB < 0)
    fprintf(stderr, "Imodv warning: no single buffer visual available.\n");
  else if (Imod_debug)
    printf("SB visual: %d color bits, %d depth bits, stereo %d\n",
	   colorSB, depthSB, a->stereoSB);

  // set to double buffer if visual exists
  a->db = depthDB >= 0 ? 1 : 0;
  return 0;
}

// OPEN THE MAIN WINDOW
static int openWindow(ImodvApp *a)
{
  int deskWidth = QApplication::desktop()->width();
  int deskHeight = QApplication::desktop()->height();
  int newWidth = a->winx;
  int needy = a->winy;

  a->lighting = Imodv->imod->view->world & VIEW_WORLD_LIGHT;
  a->lowres = Imodv->imod->view->world & VIEW_WORLD_LOWRES;
  a->mainWin = new ImodvWindow(a->standalone, a->enableDepthDB, 
                               a->enableDepthSB, a->lighting, a->lowres);

  if (!a->mainWin)
    return 1;

  ImodvClosed = 0;
  imodvSetCaption();

  // If the user did not enter a window size, just resize to that before
  // showing
  if (a->winx == DEFAULT_XSIZE && a->winy == DEFAULT_YSIZE) {
    a->mainWin->resize(a->winx, a->winy);
    if (a->fullscreen)
      a->mainWin->showMaximized();
    a->mainWin->show();
  } else {

    // Otherwise, show the window and then set the size and fix the position
    // This seems not to cause two initial draws, but it might someday
    a->mainWin->show();
    int newHeight = needy + a->mainWin->height() - a->winy;
    if (newHeight > deskHeight - 50)
      newHeight = deskHeight - 50;
    if (newWidth > deskWidth - 30)
      newWidth = deskWidth -30;

    QPoint pos = a->mainWin->pos();
    int xleft = pos.x();
    int ytop = pos.y();
    if (xleft + newWidth > deskWidth - 16)
      xleft = deskWidth - 16 - newWidth;
    if (ytop + newHeight > deskHeight - 40)
      ytop = deskHeight - 40 - newHeight;

    if (Imod_debug)
      fprintf(stderr, "Sizes: imodv %d %d, GL %d %d: "
            "resize %d %d\n", a->mainWin->width(), a->mainWin->height(),
             a->mainWin->mCurGLw->width(), 
            a->mainWin->mCurGLw->height(),
            newWidth, newHeight);
    a->mainWin->setGeometry(xleft, ytop, newWidth, newHeight);
  }

  return(0);
}

// OPEN THE MODELS IN IMODV
static int load_models(int n, char **fname, ImodvApp *a)
{
  int i, ob, co;

  if (n < 1)
    return(0);
  a->mod = (Imod **)malloc(sizeof(Imod *) * n);
  a->nm = n;
  a->cm = 0;
  for(i = 0; i < n; i++){
    a->mod[i] = imodRead((char *)(QDir::convertSeparators(QString(fname[i]))).
      latin1());
    if (!a->mod[i]){
      fprintf(stderr, "Error loading %s\n", fname[i]);
      return(-1);
    }

    /* DNM 6/20/01: find out max time and set current time */
    a->mod[i]->tmax = 0;
    for (ob = 0; ob < a->mod[i]->objsize; ob++)
      for (co = 0; co < a->mod[i]->obj[ob].contsize; co++)
        if (a->mod[i]->tmax < a->mod[i]->obj[ob].cont[co].type)
          a->mod[i]->tmax = a->mod[i]->obj[ob].cont[co].type;
    a->mod[i]->ctime = a->mod[i]->tmax ? 1 : 0;

    /* DNM: changes for storage of object properties in view and 
       relying on default scaling.  Also, make sure every model has
       the view to use set up */

    imodViewStore(a->mod[i], 0);
    if (!a->mod[i]->cview){
      imodViewModelDefault(a->mod[i], a->mod[i]->view);
    } else
      imodViewUse(a->mod[i]);
  }
  a->imod = (a->mod[a->cm]);
  /* DNM 8/3/01: start with current object if defined */
  if (a->imod->cindex.object >= 0 && 
      a->imod->cindex.object < a->imod->objsize) {
    a->ob = a->imod->cindex.object;
    a->obj = &(a->imod->obj[a->ob]);
  }
   
  return(0);
}

// THE ENTRY POINT FOR STANDALONE IMODV
int imodv_main(int argc, char **argv, char *cmdLineStyle)
{
  int i;
  ImodvApp *a = Imodv;
  a->standalone = 1;
  imodv_init(a, &Imodv_mdraw);

  //  open_display(&argc, argv, Imodv);
  QApplication myapp(argc, argv);

  ImodPrefs = new ImodPreferences(cmdLineStyle);

  // Parse options
  for (i = 1; i < myapp.argc(); i++){
    if (myapp.argv()[i][0] == '-'){
      switch (myapp.argv()[i][1]){

      case 'b':
	a->rbgname = strdup(myapp.argv()[++i]);
	break;

      case 'D':
	Imod_debug = 1;
	break;

      case 'f':
	a->fullscreen = 1;
	break;

      case 's':
	sscanf(myapp.argv()[++i], "%d%*c%d", &a->winx, &a->winy);
	break;

      default:
        if (strcmp("-imodv", myapp.argv()[i]) && 
            strcmp("-view", myapp.argv()[i])) {
          fprintf(stderr, "Imodv error: illegal option %s\n", myapp.argv()[i]);
          exit(1);
        }

      }
    } else
      break;
  }

  a->db        = 1;

  // Make a color from the named color; fallback to black
  QString qstr = a->rbgname;
  a->rbgcolor->setNamedColor(a->rbgname);
  if (!a->rbgcolor->isValid())
    a->rbgcolor->setRgb(0, 0, 0);

  if (getVisuals(a) != 0) {
    fprintf(stderr, "imodv error: Couldn't get rendering visual.\n");
    exit(-1);
  }

  if (myapp.argc() - i < 1)
    usage(myapp.argv()[0]);

  /* DNM 1/29/03: already forked in imod, no need to do it here */
  
  if (load_models(myapp.argc() - i, &(myapp.argv()[i]), Imodv))
    exit(-1);

  openWindow(Imodv);

  diaSetTitle("Imodv");

  return myapp.exec();
}


// THE CALL TO OPEN THE MODEL VIEW WINDOW FROM IMOD
void imodv_open()
{
  ImodView *vw = App->cvi;
  Imodv = &ImodvStruct;
  ImodvStruct.md = &Imodv_mdraw;
  ImodvApp *a = Imodv;

  int ob, co;
  Imod *imod = vw->imod;
  int hasPoints = 0;

  /* mt model ? */
  if (!imod){
    wprint("Model View didn't open because "
           "there is no model loaded.\n");
    return;
  }
  for(ob = 0; ob < imod->objsize; ob++){
    if (hasPoints > 1) break;
    for(co = 0; co < imod->obj[ob].contsize; co++){
      hasPoints += imod->obj[ob].cont[co].psize;
      if (hasPoints > 1) break;
    }
  }
  if (hasPoints < 2){
    wprint("Model View didn't open because model has no points.\n");
    return;
  }

  /* check for already open? */
  if (!ImodvClosed){
    a->mainWin->raise();
    return;
  }

  initstruct(vw, a);

  if (getVisuals(a) != 0) {
    wprint("Couldn't get rendering visual for model view."
           "  Try running imodv separately.\n");
    imodMatDelete(a->mat);
    imodMatDelete(a->rmat);
    return;
  }

  if (openWindow(a)) {
    wprint("Failed to open model view window window.\n");
    return;
  }
    
}

// TO CLOSE FROM IMOD
void imodv_close()
{
  if (!ImodvClosed)
    Imodv->mainWin->close();
}

// TO DRAW THE MODEL FROM IMOD
void imodv_draw()
{
  if (!ImodvClosed)
    imodvDraw(Imodv);
  return;
}

/* DNM: a routine for imod to notify imodv of a change in model */
void imodv_new_model(Imod *mod)
{
  if (ImodvClosed)
    return;
  Imodv->imod = mod;
  Imodv->mod[0] = mod;

  /* Set up the views and scaling, notify everybody of changes */
  imodViewStore(mod, 0);
     
  if (!mod->cview){
    imodViewModelDefault(mod, mod->view);
  }else {
    imodViewUse(mod);
    imodvDrawImodImages();
  }
  imodvSelectModel(Imodv, 0);
}

void imodvSetCaption()
{
  ImodvApp *a = Imodv;
  char *window_name;
  QString str;
  if (ImodvClosed)
    return;

  window_name = imodwEithername((char *)(a->standalone ? "Imodv:" : 
                                 "Imod Model View: "), a->imod->fileName, 1);
  if (window_name) {
    str = window_name;
    free(window_name);
  } 
  if (str.isEmpty())
    str = "Imod Model View";

  a->mainWin->setCaption(str);
}

// To call imodDraw if not in standalone mode
void imodvDrawImodImages()
{
  if (Imodv->standalone)
    return;
  imodDraw(App->cvi, IMOD_DRAW_MOD);
  imod_object_edit_draw();
}

/*
$Log$
Revision 4.4  2003/03/26 23:22:00  mast
Set up to use preferences

Revision 4.3  2003/03/04 21:41:05  mast
Added function for refreshing imod windows from imodv

Revision 4.2  2003/02/27 17:42:38  mast
Remove include of unistd for windows

Revision 4.1  2003/02/10 20:29:00  mast
autox.cpp

Revision 1.1.2.11  2003/01/29 17:50:58  mast
Fork now happens before imodv_main is called

Revision 1.1.2.10  2003/01/29 01:29:42  mast
add call for imod to close imodv

Revision 1.1.2.9  2003/01/27 00:30:07  mast
Pure Qt version and general cleanup

Revision 1.1.2.8  2003/01/01 19:12:31  mast
changes to start Qt application in standalone mode

Revision 1.1.2.7  2003/01/01 05:46:29  mast
changes for qt version of stereo

Revision 1.1.2.6  2002/12/18 04:15:14  mast
new includes for imodv modules

Revision 1.1.2.5  2002/12/17 22:28:20  mast
cleanup of unused variables and SGI errors

Revision 1.1.2.4  2002/12/17 18:42:22  mast
Qt version, incorporating ximodv startup code

Revision 1.1.2.3  2002/12/14 05:41:08  mast
Got qxt startup in the right place

Revision 1.1.2.2  2002/12/06 21:58:40  mast
*** empty log message ***

Revision 1.1.2.1  2002/12/05 16:28:37  mast
Open a qxt application

Revision 3.5  2002/12/01 15:34:41  mast
Changes to get clean compilation with g++

Revision 3.4  2002/11/30 06:07:22  mast
Corrected tables after addition of true-15 visual

Revision 3.3  2002/11/27 03:29:45  mast
Made it look for both single and double buffer visuals as long as they
are both RGB or both color index.  Added a true 15 visual, better than
pseudo 12 (present on O2).

Revision 3.2  2002/09/04 00:24:48  mast
Added CVS header.  Changed to getting visuals then passing them to GLw.

*/
